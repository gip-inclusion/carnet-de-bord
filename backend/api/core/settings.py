import os

from pydantic import BaseSettings

dir_path = os.path.dirname(os.path.realpath(__file__))


class Settings(BaseSettings):
    app_url: str
    database_url: str
    hasura_graphql_jwt_secret: str

    smtp_host: str
    smtp_port: str
    smtp_user: str | None
    smtp_pass: str | None

    min_pool_size: int = 2
    max_pool_size: int = 20

    PE_AUTH_BASE_URL: str = "https://entreprise.pole-emploi.fr"
    PE_BASE_URL: str = "https://api.emploi-store.fr"
    PE_SCOPE = "api_referentielagencesv1 organisationpe"
    PE_CLIENT_ID: str = "CLIENT_ID"
    PE_CLIENT_SECRET: str = "CLIENT_SECRET"

    V1_PREFIX: str = "/v1"
    MAIL_FROM: str = "support.carnet-de-bord@fabrique.social.gouv.fr"
    LOG_FORMAT = "[%(asctime)s:%(filename)s:%(lineno)s - %(funcName)20s() ] %(message)s"

    class Config:
        env_file = ".env"
        env_file_encoding = "utf-8"


# By default, load the file of the parent project if the ENV_FILE
# variable doesn't exist
#
# If you want to specify a specific env file, you can setup
# the ENV_FILE variable value
#
# Something like that: ENV_FILE="../.env" python scripts/connect_to_db.py

env_file_path = os.getenv(
    "ENV_FILE", os.path.join(dir_path, "..", "..", "..", "app", ".env")
)


if os.path.exists(env_file_path):
    settings = Settings(_env_file=env_file_path, _env_file_encoding="utf-8")
else:
    settings = Settings()
