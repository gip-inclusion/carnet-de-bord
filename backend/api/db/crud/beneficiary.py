from datetime import date
from typing import List
from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection
from asyncpg.exceptions import UniqueViolationError

from api.db.crud.notebook import (
    NOTEBOOK_BASE_FIELDS,
    NOTEBOOK_BASE_JOINS,
    parse_notebook_from_record,
)
from api.db.models.beneficiary import Beneficiary, BeneficiaryImport

BENEFICIARY_BASE_QUERY = (
    """SELECT b.*, acc.id as account_id,"""
    + NOTEBOOK_BASE_FIELDS
    + """
    FROM public.beneficiary b
    LEFT JOIN notebook n ON n.beneficiary_id = b.id
    """
    + NOTEBOOK_BASE_JOINS
    + """
    LEFT JOIN public.account acc
    ON acc.beneficiary_id = b.id
"""
)


# see https://betakuang.medium.com/why-postgresqls-on-conflict-cannot-find-my-partial-unique-index-552327b85e1
firstname_lastname_date_of_birth_unique_idx = (
    "(LOWER(trim(firstname)), LOWER(trim(lastname)), date_of_birth, deployment_id)"
)


async def select_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    return await connection.fetch(
        f"""
SELECT (firstname, lastname, date_of_birth, internal_id, deployment_id, id)
FROM beneficiary
WHERE (firstname = $1 AND lastname = $2 AND date_of_birth = $3)
OR (internal_id = $4 AND deployment_id = $5)
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.date_of_birth,
        beneficiary.si_id,
        deployment_id,
    )


async def update_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
UPDATE beneficiary SET mobile_number = $1
where internal_id = $2
returning id
        """,
        beneficiary.phone_number,
        beneficiary.si_id,
    )


async def insert_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
INSERT INTO BENEFICIARY (firstname, lastname, internal_id, date_of_birth, deployment_id, mobile_number)
values ($1, $2, $3, $4, $5, $6)
returning id
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.si_id,
        beneficiary.date_of_birth,
        deployment_id,
        beneficiary.phone_number,
    )


async def import_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    result = select_beneficiary(connection, beneficiary, deployment_id)
    existing_rows = await result
    if len(existing_rows) == 0:
        print("inserting new user")
        await insert_beneficiary(connection, beneficiary, deployment_id)
    else:
        print(existing_rows[0][0][4])
        if (
            len(existing_rows) == 1
            and existing_rows[0][0][0] == beneficiary.firstname
            and existing_rows[0][0][1] == beneficiary.lastname
            and existing_rows[0][0][2] == beneficiary.date_of_birth
            and existing_rows[0][0][3] == beneficiary.si_id
        ):
            # Doesn't work somehow
            #     and existing_rows[0][0][4] == deployment_id
            print("updating existing user")
            await update_beneficiary(connection, beneficiary, deployment_id)
        else:
            print("data problem")


async def get_beneficiary_with_query(
    connection: Connection, query: str, *args
) -> Beneficiary | None:

    async with connection.transaction():

        beneficiary = None

        beneficiary_records: List[Record] = await connection.fetch(
            BENEFICIARY_BASE_QUERY + query,
            *args,
        )

        for beneficiary_record in beneficiary_records:

            if beneficiary is None:
                beneficiary = Beneficiary.parse_obj(beneficiary_record)

            if beneficiary_record["n_id"] is not None:
                beneficiary.notebook = await parse_notebook_from_record(
                    beneficiary_record,
                    notebook=beneficiary.notebook,
                )

        return beneficiary


async def get_beneficiary_from_personal_information(
    connection: Connection, firstname: str, lastname: str, birth_date: date
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection,
        # We use LOWER(BTRIM(field)) here to match the index so postgres
        # can use the index rather than scanning the whole table
        "WHERE LOWER(BTRIM(b.firstname)) = LOWER($1) AND "
        "LOWER(BTRIM(b.lastname)) = LOWER($2) AND "
        "b.date_of_birth = $3",
        firstname,
        lastname,
        birth_date,
    )


async def get_beneficiary_by_id(
    connection: Connection, beneficiary_id: UUID
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection, "WHERE b.id = $1", beneficiary_id
    )
