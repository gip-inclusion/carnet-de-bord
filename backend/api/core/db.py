import asyncpg
from asyncpg.pool import Pool

from api.core.settings import settings


async def get_connection_pool(database_url: str) -> Pool | None:
    return await asyncpg.create_pool(
        database_url, min_size=settings.min_pool_size, max_size=settings.max_pool_size
    )
