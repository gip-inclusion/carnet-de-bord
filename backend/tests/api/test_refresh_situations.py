import json
from datetime import datetime, timedelta, timezone
from urllib.parse import quote
from uuid import UUID

import httpx
import pytest
import respx
from asyncpg import Connection, Record
from gql import gql

from cdb.api.core.settings import Settings, settings
from cdb.api.db.crud.external_data import (
    insert_external_data,
    insert_external_data_info,
)
from cdb.api.db.graphql import get_client
from cdb.api.db.models.beneficiary import Beneficiary
from cdb.api.db.models.external_data import (
    ExternalDataInfoInsert,
    ExternalDataInsert,
    ExternalSource,
)
from cdb.api.db.models.notebook import Notebook
from tests.mocks.pole_emploi_recherche_usagers import (
    PE_API_RECHERCHE_USAGERS_RESULT_KO_MOCK,
    PE_API_RECHERCHE_USAGERS_RESULT_OK_MOCK,
)
from tests.utils.approvaltests import verify_as_json

pytestmark = pytest.mark.graphql


@pytest.fixture
async def pe_settings():
    settings.PE_CLIENT_SECRET = "pe_client_secret"
    settings.PE_CLIENT_ID = "pe_client_id"
    settings.ENABLE_SYNC_CONTRAINTES = True
    yield settings


async def test_produces_a_401_when_the_token_is_missing(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
):
    response = await test_client.post(
        "/v1/notebooks/refresh-situations-from-pole-emploi",
        json=build_payload(notebook_id=beneficiary_sophie_tifour.notebook.id),
    )
    assert response.status_code == 401
    json = response.json()
    assert json["detail"] == "Missing credentials"


@pytest.mark.graphql
@respx.mock
async def test_nominal(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    sophie_tifour_pe_diagnostic_with_contraintes: dict,
    pe_settings,
):
    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        sophie_tifour_pe_diagnostic_with_contraintes,
    )

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    expect_ok_response(
        response,
        {
            "data_has_been_updated": True,
            "external_data_has_been_updated": True,
            "has_pe_diagnostic": True,
        },
    )

    situations = await get_notebook_situations_by_id(notebook_sophie_tifour.id)
    assert situations

    assert [
        (
            situation["refSituation"]["description"],
            situation["createdAt"].split("T")[0],
        )
        for situation in situations
    ] == [
        ("Doit quitter le logement", "2023-05-12"),
        ("Hébergé chez un tiers", "2021-09-21"),
    ]


@pytest.mark.graphql
@respx.mock
async def test_refreshes_on_new_data_from_pe(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    sophie_tifour_pe_diagnostic_with_add_only_contraintes: dict,
    notebook_sophie_tifour: Notebook,
    pe_settings,
):
    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        sophie_tifour_pe_diagnostic_with_add_only_contraintes,
    )

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    expect_ok_response(
        response,
        {
            "data_has_been_updated": True,
            "external_data_has_been_updated": True,
            "has_pe_diagnostic": True,
        },
    )


@pytest.mark.graphql
@respx.mock
# updates cdb data and external data when there are new constraints and no external data
async def test_updates_cdb_data_and_external_data_when_there_are_new_constraintes_and_no_external_data(  # noqa: E501
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    sophie_tifour_pe_diagnostic_with_add_only_contraintes: dict,
    pe_settings,
):
    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        sophie_tifour_pe_diagnostic_with_add_only_contraintes,
    )

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    expect_ok_response(
        response,
        {
            "data_has_been_updated": True,
            "external_data_has_been_updated": True,
            "has_pe_diagnostic": True,
        },
    )


@pytest.mark.graphql
@respx.mock
async def test_does_nothing_when_pe_has_no_info_for_our_beneficiary(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    pe_settings: Settings,
):
    mock_pe_api_not_found_individu(pe_settings, beneficiary_sophie_tifour)

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    expect_ok_response(
        response,
        {
            "data_has_been_updated": False,
            "external_data_has_been_updated": False,
            "has_pe_diagnostic": False,
        },
    )


@pytest.mark.graphql
@respx.mock
async def test_does_nothing_when_our_pe_data_was_fetched_recently(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    db_connection: Connection,
    pe_settings: Settings,
):
    external_data = await insert_external_data(
        db_connection,
        ExternalDataInsert(data={}, hash="hash", source=ExternalSource.PE_IO),
    )
    assert external_data
    await insert_external_data_info(
        db_connection,
        ExternalDataInfoInsert(
            external_data_id=external_data.id,
            beneficiary_id=beneficiary_sophie_tifour.id,
            created_at=datetime.now(tz=timezone.utc) - timedelta(minutes=59),
        ),
    )

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)
    pe_internal_id = PE_API_RECHERCHE_USAGERS_RESULT_OK_MOCK["identifiant"]
    route = respx.get(
        f"{pe_settings.PE_BASE_URL}/partenaire/diagnosticargumente/v1/individus/{quote(pe_internal_id)}",
    )
    assert not route.called

    expect_ok_response(
        response,
        {
            "data_has_been_updated": False,
            "external_data_has_been_updated": False,
            "has_pe_diagnostic": True,
        },
    )


@pytest.mark.graphql
@respx.mock
async def test_does_nothing_when_we_do_not_find_requested_notebook(
    test_client: httpx.AsyncClient,
):
    response = await call_refresh_api(
        UUID("8341681f-aef3-4d28-baf6-763435b253d6"), test_client
    )

    assert response.status_code == 200
    assert response.json() is None


@pytest.mark.graphql
@respx.mock
async def test_refreshes_data_from_pe_when_they_have_expired(
    test_client: httpx.AsyncClient,
    db_connection: Connection,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    sophie_tifour_pe_diagnostic_with_contraintes: dict,
    pe_settings: Settings,
):
    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        sophie_tifour_pe_diagnostic_with_contraintes,
    )

    await given_beneficiary_has_external_data_at_dates(
        [expired_date()], beneficiary_sophie_tifour, db_connection
    )

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    expect_ok_response(
        response,
        {
            "data_has_been_updated": True,
            "external_data_has_been_updated": True,
            "has_pe_diagnostic": True,
        },
    )


@pytest.mark.graphql
@respx.mock
async def test_does_nothing_when_at_least_one_external_data_is_fresh(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    db_connection: Connection,
    pe_settings: Settings,
):
    await given_beneficiary_has_external_data_at_dates(
        [fresh_creation_date(), expired_date()],
        beneficiary_sophie_tifour,
        db_connection,
    )
    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    pe_internal_id = PE_API_RECHERCHE_USAGERS_RESULT_OK_MOCK["identifiant"]
    route = respx.get(
        f"{pe_settings.PE_BASE_URL}/partenaire/diagnosticargumente/v1/individus/{quote(pe_internal_id)}",
    )
    assert not route.called

    expect_ok_response(
        response,
        {
            "data_has_been_updated": False,
            "external_data_has_been_updated": False,
            "has_pe_diagnostic": True,
        },
    )


@pytest.mark.graphql
@respx.mock
async def test_does_nothing_when_feature_is_disabled(
    test_client: httpx.AsyncClient,
    notebook_sophie_tifour: Notebook,
    pe_settings: Settings,
):
    pe_settings.ENABLE_SITUATION_API = False
    pe_internal_id = PE_API_RECHERCHE_USAGERS_RESULT_OK_MOCK["identifiant"]
    route = respx.get(
        f"{pe_settings.PE_BASE_URL}/partenaire/diagnosticargumente/v1/individus/{quote(pe_internal_id)}",
    )
    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)
    assert not route.called
    expect_ok_response(
        response,
        {
            "data_has_been_updated": False,
            "external_data_has_been_updated": False,
            "has_pe_diagnostic": False,
        },
    )

    pe_settings.ENABLE_SITUATION_API = True


@pytest.mark.graphql
@respx.mock
async def test_saves_the_new_pe_diagnostic_into_external_data(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    sophie_tifour_pe_diagnostic_with_add_only_contraintes: dict,
    db_connection: Connection,
    pe_settings,
):
    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        sophie_tifour_pe_diagnostic_with_add_only_contraintes,
    )
    await given_beneficiary_has_external_data_at_dates(
        [], beneficiary_sophie_tifour, db_connection
    )

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    record: Record = await get_last_saved_external_data(
        beneficiary_sophie_tifour.id, ExternalSource.PE_IO, db_connection
    )
    verify_as_json(
        {
            "data": json.loads(record["data"]),
            "source": record["source"],
            "hash": record["hash"],
        }
    )
    expect_ok_response(
        response,
        {
            "data_has_been_updated": True,
            "external_data_has_been_updated": True,
            "has_pe_diagnostic": True,
        },
    )


@pytest.mark.graphql
@respx.mock
# TODO : Changer le setup pour pas appeler 2 fois le endpoint
# TODO : On souhaite conserver la date de derniere récupération pour ne pas rappeler
# PE a chaque fois dans le cas ou les données n'ont pas changée
async def test_does_not_store_anything_if_pe_data_has_not_changed_since_last_fetch(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    sophie_tifour_pe_diagnostic_with_contraintes: dict,
    db_connection: Connection,
    pe_settings,
):
    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        sophie_tifour_pe_diagnostic_with_contraintes,
    )
    before = expired_date()

    await call_refresh_api(notebook_sophie_tifour.id, test_client)
    # update created_at so it happened in the past
    await update_external_data_created_at(
        beneficiary_sophie_tifour.id, before, db_connection
    )

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    result = await get_last_saved_external_data(
        beneficiary_sophie_tifour.id, ExternalSource.PE_IO, db_connection
    )
    assert result["created_at"] == before
    expect_ok_response(
        response,
        {
            "data_has_been_updated": False,
            "external_data_has_been_updated": False,
            "has_pe_diagnostic": True,
        },
    )


@pytest.mark.graphql
@respx.mock
async def test_does_nothing_when_data_from_pe_is_new_but_matches_our_situations(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    sophie_tifour_pe_diagnostic_with_no_new_contraintes: dict,
    pe_settings,
):
    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        sophie_tifour_pe_diagnostic_with_no_new_contraintes,
    )

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    expect_ok_response(
        response,
        {
            "data_has_been_updated": False,
            "external_data_has_been_updated": True,
            "has_pe_diagnostic": True,
        },
    )


@pytest.mark.graphql
@respx.mock
async def test_update_situation_when_received_situation_are_an_empty_array(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    sophie_tifour_pe_diagnostic_with_empty_contraintes: dict,
    pe_settings,
):
    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        sophie_tifour_pe_diagnostic_with_empty_contraintes,
    )

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    expect_ok_response(
        response,
        {
            "data_has_been_updated": True,
            "external_data_has_been_updated": True,
            "has_pe_diagnostic": True,
        },
    )

    situations = await get_notebook_situations_by_id(notebook_sophie_tifour.id)
    assert len(situations) == 0


# TODO: remove ?
@pytest.mark.graphql
async def test_does_nothing_when_the_pe_client_id_is_missing(
    test_client: httpx.AsyncClient,
    notebook_sophie_tifour: Notebook,
    pe_settings,
):
    pe_settings.PE_CLIENT_ID = ""

    response = await test_client.post(
        "/v1/notebooks/refresh-situations-from-pole-emploi",
        headers={"secret-token": "action_secret_token"},
        json=build_payload(notebook_id=notebook_sophie_tifour.id),
    )

    expect_ok_response(
        response,
        {
            "data_has_been_updated": False,
            "external_data_has_been_updated": False,
            "has_pe_diagnostic": False,
        },
    )


# TODO: remove ?
@pytest.mark.graphql
async def test_does_nothing_when_the_pe_client_secret_is_missing(
    test_client: httpx.AsyncClient,
    notebook_sophie_tifour: Notebook,
    pe_settings,
):
    pe_settings.PE_CLIENT_SECRET = ""

    response = await test_client.post(
        "/v1/notebooks/refresh-situations-from-pole-emploi",
        headers={"secret-token": "action_secret_token"},
        json=build_payload(notebook_id=notebook_sophie_tifour.id),
    )

    expect_ok_response(
        response,
        {
            "data_has_been_updated": False,
            "external_data_has_been_updated": False,
            "has_pe_diagnostic": False,
        },
    )


@pytest.mark.graphql
@respx.mock
async def test_produces_a_400_when_the_pe_api_returns_a_500(
    test_client: httpx.AsyncClient,
    notebook_sophie_tifour: Notebook,
    pe_settings,
):
    """To forward api errors through the Hasura action it needs a 400 error"""
    respx.post(f"{pe_settings.PE_AUTH_BASE_URL}/connexion/oauth2/access_token").mock(
        return_value=httpx.Response(
            500,
            json={
                "token_type": "foo",
                "access_token": "batman",
                "expires_in": 3600,
            },
        )
    )
    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    assert response.status_code == 400


@pytest.mark.graphql
@respx.mock
async def test_does_not_update_cdb_when_sync_flag_is_false(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    pe_settings: Settings,
    sophie_tifour_pe_diagnostic_with_empty_contraintes,
    notebook_sophie_tifour: Notebook,
):
    pe_settings.ENABLE_SYNC_CONTRAINTES = False

    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        sophie_tifour_pe_diagnostic_with_empty_contraintes,
    )

    response = await call_refresh_api(notebook_sophie_tifour.id, test_client)

    expect_ok_response(
        response,
        {
            "data_has_been_updated": False,
            "external_data_has_been_updated": True,
            "has_pe_diagnostic": True,
        },
    )

    situations = await get_notebook_situations_by_id(notebook_sophie_tifour.id)
    assert len(situations) != 0


async def get_last_saved_external_data(
    beneficiary__id: UUID, source: ExternalSource, db_connection
) -> Record:
    result = await db_connection.fetchrow(
        """
        SELECT data, source, hash, edi.created_at
        FROM external_data
        LEFT JOIN external_data_info edi ON external_data_id = id
        WHERE beneficiary_id = $1 AND source = $2
        ORDER BY edi.created_at DESC
        """,
        beneficiary__id,
        source,
    )
    return result


async def update_external_data_created_at(
    beneficiary_id: UUID, created_at: datetime, db_connection
):
    await db_connection.fetchrow(
        """
        UPDATE external_data_info set created_at=$1
        WHERE beneficiary_id = $2
        """,
        created_at,
        beneficiary_id,
    )


async def add_situation_to_notebook(
    notebook_id: UUID, situation_id: UUID, db_connection
):
    await db_connection.fetchrow(
        """
        INSERT INTO notebook_situation(notebook_id, situation_id) VALUES($1, $2);
        """,
        notebook_id,
        situation_id,
    )


def build_payload(notebook_id: UUID):
    return {
        "action": {"name": "refresh-situations-from-pole-emploi"},
        "request_query": "",
        "session_variables": {},
        "input": {
            "notebookId": notebook_id,
        },
    }


async def get_notebook_situations_by_id(id):
    async with get_client.gql_client_backend_only() as session:
        result = await session.execute(
            gql(
                """
            query situations($id: uuid!) {
                notebook_situation(
                    where: {notebookId: {_eq: $id}  deletedAt: {_is_null: true}}
                    order_by: {createdAt: desc}
                ) {
                    id, refSituation{ description} createdAt
                }
            }
        """
            ),
            variable_values={"id": id},
        )
        return result["notebook_situation"]


def mock_pe_api(
    settings: Settings,
    beneficiary: Beneficiary,
    pe_diagnostic: dict | None,
):
    respx.post(f"{settings.PE_AUTH_BASE_URL}/connexion/oauth2/access_token").mock(
        return_value=httpx.Response(
            200,
            json={
                "token_type": "foo",
                "access_token": "batman",
                "expires_in": 3600,
            },
        )
    )

    respx.post(
        f"{settings.PE_BASE_URL}/partenaire/rechercher-usager/v1/usagers/recherche",
        json={
            "nir": beneficiary.nir,
            "dateNaissance": beneficiary.date_of_birth.strftime("%Y-%m-%d"),
        },
    ).mock(
        return_value=httpx.Response(200, json=PE_API_RECHERCHE_USAGERS_RESULT_OK_MOCK)
    )

    pe_internal_id = PE_API_RECHERCHE_USAGERS_RESULT_OK_MOCK["identifiant"]

    if not pe_diagnostic:
        dossier_individu_response_code = 204
        dossier_individu_response = None
    else:
        dossier_individu_response_code = 200
        dossier_individu_response = pe_diagnostic

    respx.get(
        f"{settings.PE_BASE_URL}/partenaire/diagnosticargumente/v1/individus/{quote(pe_internal_id)}",
    ).mock(
        return_value=httpx.Response(
            dossier_individu_response_code, json=dossier_individu_response
        )
    )


def mock_pe_api_not_found_individu(
    settings: Settings,
    beneficiary: Beneficiary,
):
    respx.post(f"{settings.PE_AUTH_BASE_URL}/connexion/oauth2/access_token").mock(
        return_value=httpx.Response(
            200,
            json={
                "token_type": "foo",
                "access_token": "batman",
                "expires_in": 3600,
            },
        )
    )

    respx.post(
        f"{settings.PE_BASE_URL}/partenaire/rechercher-usager/v1/usagers/recherche",
        json={
            "nir": beneficiary.nir,
            "dateNaissance": beneficiary.date_of_birth.strftime("%Y-%m-%d"),
        },
    ).mock(
        return_value=httpx.Response(200, json=PE_API_RECHERCHE_USAGERS_RESULT_KO_MOCK)
    )


@pytest.fixture(scope="session")
def sophie_tifour_pe_diagnostic_with_empty_contraintes() -> dict:
    return {
        "besoinsParDiagnosticIndividuDtos": [],
        "contraintesIndividusDto": {
            "conseiller": "TNAN0260",
            "dateDeModification": "2023-05-12T12:54:39.000+00:00",
            "code": "7",
            "libelle": "Résoudre ses contraintes personnelles",
            "contraintes": [],
        },
    }


@pytest.fixture(scope="session")
def sophie_tifour_pe_diagnostic_with_contraintes() -> dict:
    return {
        "besoinsParDiagnosticIndividuDtos": [],
        "contraintesIndividusDto": {
            "conseiller": "TNAN0260",
            "dateDeModification": "2023-05-12T12:54:39.000+00:00",
            "code": "7",
            "libelle": "Résoudre ses contraintes personnelles",
            "contraintes": [
                {
                    "code": "27",
                    "libelle": "Faire face à des difficultés de logement",
                    "valeur": "OUI",
                    "date": "2023-05-12T12:54:39.000+00:00",
                    "situations": [
                        {
                            "code": "24",
                            "libelle": "Hébergé chez un tiers",
                            "valeur": "OUI",
                        },
                        {
                            "code": "27",
                            "libelle": "Doit quitter le logement",
                            "valeur": "OUI",
                        },
                    ],
                    "objectifs": [],
                },
                {
                    "code": "29",
                    "libelle": "Faire face à des difficultés administratives ou juridiques",  # noqa: E501
                    "valeur": "CLOTUREE",
                    "date": None,
                    "situations": [
                        {
                            "code": "36",
                            "libelle": "Besoin d'etre guidé dans le cadre d'un accès aux droits",  # noqa: E501
                            "valeur": "CLOTUREE",
                        },
                    ],
                    "objectifs": [],
                },
            ],
        },
    }


@pytest.fixture(scope="session")
def sophie_tifour_pe_diagnostic_with_no_new_contraintes() -> dict:
    return {
        "besoinsParDiagnosticIndividuDtos": [],
        "contraintesIndividusDto": {
            "conseiller": "TNAN0260",
            "dateDeModification": "2023-05-12T12:54:39.000+00:00",
            "code": "7",
            "libelle": "Résoudre ses contraintes personnelles",
            "contraintes": [
                {
                    "code": "27",
                    "libelle": "Faire face à des difficultés de logement",
                    "valeur": "OUI",
                    "date": "2023-05-12T12:54:39.000+00:00",
                    "situations": [
                        {
                            "code": "24",
                            "libelle": "Hébergé chez un tiers",
                            "valeur": "OUI",
                        },
                        {
                            "code": "27",
                            "libelle": "Besoin d'être guidé dans le cadre d'un accès aux droits",  # noqa: E501
                            "valeur": "OUI",
                        },
                        {
                            "code": "26",
                            "libelle": "Prêt  à suivre une formation",
                            "valeur": "OUI",
                        },
                    ],
                    "objectifs": [],
                },
            ],
        },
    }


@pytest.fixture(scope="session")
def sophie_tifour_pe_diagnostic_with_add_only_contraintes() -> dict:
    return {
        "besoinsParDiagnosticIndividuDtos": [],
        "contraintesIndividusDto": {
            "conseiller": "TNAN0260",
            "dateDeModification": "2023-05-12T12:54:39.000+00:00",
            "code": "7",
            "libelle": "Résoudre ses contraintes personnelles",
            "contraintes": [
                {
                    "code": "27",
                    "libelle": "Faire face à des difficultés de logement",
                    "valeur": "OUI",
                    "date": "2023-05-12T12:54:39.000+00:00",
                    "situations": [
                        {
                            "code": "24",
                            "libelle": "Hébergé chez un tiers",
                            "valeur": "OUI",
                        },
                        {
                            "code": "27",
                            "libelle": "Besoin d'être guidé dans le cadre d'un accès aux droits",  # noqa: E501
                            "valeur": "OUI",
                        },
                        {
                            "code": "26",
                            "libelle": "Prêt  à suivre une formation",
                            "valeur": "OUI",
                        },
                    ],
                    "objectifs": [],
                },
                {
                    "code": "29",
                    "libelle": "Faire face à des difficultés administratives ou juridiques",  # noqa: E501
                    "valeur": "OUI",
                    "date": "2022-05-12T12:54:39.000+00:00",
                    "situations": [
                        {
                            "code": "36",
                            "libelle": "Attend un enfant ou plus",
                            "valeur": "OUI",
                        },
                    ],
                    "objectifs": [],
                },
            ],
        },
    }


async def given_beneficiary_has_external_data_at_dates(
    dates, beneficiary, db_connection
):
    for date in dates:
        await given_beneficiary_has_pe_io_data(beneficiary, date, db_connection)


def expect_ok_response(response, expected_response_dict):
    assert response.status_code == 200
    assert response.json().items() == expected_response_dict.items()


async def call_refresh_api(uuid: UUID, test_client):
    response = await test_client.post(
        "/v1/notebooks/refresh-situations-from-pole-emploi",
        headers={"secret-token": "action_secret_token"},
        json=build_payload(notebook_id=uuid),
    )
    return response


def fresh_creation_date():
    return expired_date() + timedelta(minutes=1)


async def given_beneficiary_has_pe_io_data(beneficiary, created_at, db_connection):
    external_data = await insert_external_data(
        db_connection,
        ExternalDataInsert(data={}, hash="hash", source=ExternalSource.PE_IO),
    )
    assert external_data
    await insert_external_data_info(
        db_connection,
        ExternalDataInfoInsert(
            external_data_id=external_data.id,
            beneficiary_id=beneficiary.id,
            created_at=created_at,
        ),
    )


async def given_external_data_with_pe_io(db_connection):
    external_data = await insert_external_data(
        db_connection,
        ExternalDataInsert(data={}, hash="hash", source=ExternalSource.PE_IO),
    )
    assert external_data
    return external_data


def expired_date():
    return datetime.now(tz=timezone.utc) - timedelta(hours=1)
