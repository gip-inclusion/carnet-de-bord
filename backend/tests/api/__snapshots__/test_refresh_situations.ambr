# serializer version: 1
# name: test_saves_the_new_contraintes_into_external_data
  dict({
    'data': '[{"id": 27, "nom": "Faire face à des difficultés de logement", "date": "2023-05-12T12:54:39+00:00", "valeur": "NON_ABORDEE", "situations": [{"code": "24", "valeur": "OUI", "libelle": "Hébergé chez un tiers"}, {"code": "27", "valeur": "OUI", "libelle": "Besoin d\'être guidé dans le cadre d\'un accès aux droits"}, {"code": "26", "valeur": "OUI", "libelle": "Prêt  à suivre une formation"}]}, {"id": 29, "nom": "Faire face à des difficultés administratives ou juridiques", "date": "2022-05-12T12:54:39+00:00", "valeur": "CLOTUREE", "situations": [{"code": "36", "valeur": "OUI", "libelle": "Attend un enfant ou plus"}]}]',
    'hash': '14618249690d708291ade94cb220d1d28a51e0c5e52e996d90142a9e71eb14ba',
    'source': 'peio',
  })
# ---
